### PR TITLE
Implement `ImageListDesigner`

### LINKED ISSUE #9482: Cannot open smart tag window for ImageList in DemoConsole application
### .NET version

.Net 8.0 SDK: 8.0.100-preview.7.23362.29

### Did it work in .NET Framework?

Not tested/verified

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, still reproduces in .Net 7.0

### Issue description

In [DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole) test app of Winforms repo, cannot open smart tag window for ImageList to do some operation. 
image

**Expected result:**
image

### Steps to reproduce

Test sample: [DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole) in Winforms repo

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #9482
related to #4908

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9493)

### CODE REVIEW

**Comment by elachlan** [comment]
@dmitrii-drobotov I ended up implementing it.

Found the source code at: 
http://www.dotnetframework.org/default.aspx/DotNET/DotNET/8@0/untmp/whidbey/REDBITS/ndp/fx/src/Designer/WinForms/System/WinForms/Design/ImageListDesigner@cs/2/ImageListDesigner@cs

**Comment by elachlan** [comment]
@Olina-Zhang can someone from the team give it a test. I played around with it and it seems to be working.

**Comment by Olina-Zhang** [comment]
@elachlan I tested this PR, all things work well except one issue when expanding Image Bit Depth to choose a blank one, please see following gif:
exception
Call stack:
System.ComponentModel.InvalidEnumArgumentException
  HResult=0x80070057
  Message=The value of argument 'value' (0) is invalid for Enum type 'ColorDepth'. (Parameter 'value')
  Source=System.Windows.Forms
  StackTrace:
   at SourceGenerated.EnumValidator.ReportEnumValidationError(String parameterName, Int32 value, Type enumType) in C:\Users\v-olzhan\source\repos\winforms_elachlan\src\System.Windows.Forms\src\System.Windows.Forms.PrivateSourceGenerators\System.Windows.Forms.PrivateSourceGenerators.EnumValidationGenerator\Validation.cs:line 872
   at SourceGenerated.EnumValidator.Validate(ColorDepth enumToValidate, String parameterName) in C:\Users\v-olzhan\source\repos\winforms_elachlan\src\System.Windows.Forms\src\System.Windows.Forms.PrivateSourceGenerators\System.Windows.Forms.PrivateSourceGenerators.EnumValidationGenerator\Validation.cs:line 395
   at System.Windows.Forms.ImageList.set_ColorDepth(ColorDepth value) in C:\Users\v-olzhan\source\repos\winforms_elachlan\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.cs:line 90
   at System.Windows.Forms.Design.ImageListDesigner.set_ColorDepth(ColorDepth value) in C:\Users\v-olzhan\source\repos\winforms_elachlan\src\System.Windows.Forms.Design\src\System\Windows\Forms\Design\ImageListDesigner.cs:line 37
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)

**Comment by elachlan** [comment]
Awesome. Ill sort it out soon.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.cs [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+            _actionLists ??= new DesignerActionListCollection
+            {
+                new ImageListActionList(this)
+            };
+            return _actionLists;
+        }
+    }
+
+    //  Shadow ImageList.Images to allow arbitrary handle recreation.
+    internal class OriginalImageCollection : IList
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.cs:162]: You can also add Editor attribute as in original code, it's already ported.
```suggestion
    [Editor($"System.Windows.Forms.Design.ImageCollectionEditor, {AssemblyRef.SystemDesign}", typeof(UITypeEditor))]
    internal class OriginalImageCollection : IList
```

Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.cs:162]: I would also suggest to extract `OriginalImageCollection` into a separate file.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    public ImageListActionList(ImageListDesigner designer)
+        : base(designer.Component)
+    {
+        _designer = designer;
+    }
+
+    public void ChooseImages() => EditorServiceContext.EditValue(_designer, Component, "Images");
+
+    public ColorDepth? ColorDepth
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs:23]: Looks like making it nullable adds that blank row by treating null as a valid value, so I think we need to make it non-nullable.
```suggestion
    public ColorDepth ColorDepth
```

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 3]
```diff
@@ -1317,10 +1317,28 @@ Press Ctrl+Enter to accept Text.</value>
...
-    <data name="DesignerShortcutVerticalOrientation" xml:space="preserve">
+  <data name="DesignerShortcutVerticalOrientation" xml:space="preserve">
     <value>Vertical Splitter Orientation.</value>
   </data>
-    <data name="DesignerShortcutHorizontalOrientation" xml:space="preserve">
+  <data name="DesignerShortcutHorizontalOrientation" xml:space="preserve">
     <value>Horizontal Splitter Orientation.</value>
   </data>
+  <data name="ImageListActionList_ChooseImagesDescription" xml:space="preserve">
+    <value />
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:1327]:
```suggestion
    <value>Opens the Images collection editor</value>
```

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 4]
```diff
@@ -1317,10 +1317,28 @@ Press Ctrl+Enter to accept Text.</value>
...
+  </data>
+  <data name="ImageListActionList_ChooseImagesDisplayName" xml:space="preserve">
+    <value>Choose images</value>
+  </data>
+  <data name="ImageListActionList_ColorDepthDescription" xml:space="preserve">
+    <value />
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:1333]:
```suggestion
    <value>Changes the bit depth images imported into this ImageList are stored at</value>
```

File: src/System.Windows.Forms.Design/src/Resources/SR.resx [hunk 5]
```diff
@@ -1317,10 +1317,28 @@ Press Ctrl+Enter to accept Text.</value>
...
+  </data>
+  <data name="ImageListActionList_ColorDepthDisplayName" xml:space="preserve">
+    <value>Image Bit Depth</value>
+  </data>
+  <data name="ImageListActionList_ImageSizeDescription" xml:space="preserve">
+    <value />
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms.Design/src/Resources/SR.resx:1339]:
```suggestion
    <value>Changes the size images imported into this ImageList are stored at</value>
```

**Review by dmitrii-drobotov — COMMENTED** [review]
Thank you @elachlan!

File: src/System.Windows.Forms.Design/src/Resources/SR.resx (hunk 5 above)
Comment (dmitrii-drobotov) [reply src/System.Windows.Forms.Design/src/Resources/SR.resx:1339]: These are original resource values.

**Comment by dreddy-work** [comment]
related to https://github.com/dotnet/winforms/issues/4908

**Review by dreddy-work — COMMENTED** [review]
Thank you for looking into this. General principle for these out of scope designer has been to wait for external customer requests in order to understand the scenarios before proceeding with porting. Nevertheless, we have already completed the porting process for the majority of them, and we had internal discussions some time ago about porting the remaining designers to achieve completeness. I will get final confirmation here before merging this PR. Also, to make sure we are porting designers source code from trusted sources, I may ask WinForms team members to take up rest of the designers porting (they have access to private WinForms designer repo that already has .NET core source for these designers).

**Comment by elachlan** [comment]
@dreddy-work I think that is sorted. Can you have a quick look at the `AssemblyRef` changes? I moved it to common because it had shared values.

**Comment by elachlan** [comment]
@Olina-Zhang I retested your scenario and it is fixed.

**Comment by dreddy-work** [comment]
> @dreddy-work I think that is sorted. Can you have a quick look at the `AssemblyRef` changes? I moved it to common because it had shared values.

`AssemblyRef` is internal and if this move is working, it might be because of internal visible to flags. If `EnvDTE `is only used in design assembly and never in WinForms, i would keep it unchanged.

**Comment by elachlan** [comment]
> > @dreddy-work I think that is sorted. Can you have a quick look at the `AssemblyRef` changes? I moved it to common because it had shared values.
> 
> `AssemblyRef` is internal and if this move is working, it might be because of `InternVisibleTo `flags. If `EnvDTE `is only used in design assembly and never in WinForms, i would keep it unchanged.

I'll revert the change and have duplicated values in the design version.

**Comment by Olina-Zhang** [comment]
> @Olina-Zhang I retested your scenario and it is fixed.

Re-test this PR, it is fixed with no new issue found.

**Review by dmitrii-drobotov — APPROVED** [review]

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        : base(designer.Component)
+    {
+        _designer = designer;
+    }
+
+    public void ChooseImages() => EditorServiceContext.EditValue(_designer, Component, "Images");
+
+    public ColorDepth ColorDepth
+    {
+        get => (Component as ImageList)!.ColorDepth;
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs:25]:
```suggestion
        get => ((ImageList)Component).ColorDepth;
```

Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs:25]: Same in other places. Cast exception is more accurate than null reference.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.OriginalImageCollection.cs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+            return _items.IndexOf(value);
+        }
+
+        public void AddRange(ImageListImage[] values)
+        {
+            ArgumentNullException.ThrowIfNull(values);
+            foreach (ImageListImage value in values)
+            {
+                if (value is not null)
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.OriginalImageCollection.cs:115]: is this necessary with nullability enabled?

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs (hunk 6 above)
Comment (elachlan) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListActionList.cs:25]: We get a converting null value to non-null warning. So as is necessary.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.OriginalImageCollection.cs (hunk 7 above)
Comment (elachlan) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/ImageListDesigner.OriginalImageCollection.cs:115]: Yes, we should validate still.

**Comment by dreddy-work** [comment]
@elachlan, can you rebase this branch to .NET9.0-Test? We would be using it for all .NET 9.0 merges until main is open for .Net 9.0

**Comment by elachlan** [comment]
The issue this is fixing is marked for .NET 8. Can you please update that issue as well?

**Review by dreddy-work — APPROVED** [review]

**Pull request merged**